sui_programmability/framework/sources/Governance/ValidatorSet.move

```diff
@@ -0,0 +1,296 @@
+    public(friend) fun request_withdraw_stake(
+        self: &mut ValidatorSet,
+        withdraw_amount: u64,
+        ctx: &mut TxContext,
+    ) {
+        let validator_address = TxContext::sender(ctx);
+        let (found, validator_index) = find_validator(&self.pending_validators, validator_address);
+        if (found) {
+            let validator = Vector::borrow_mut(&mut self.pending_validators, validator_index);
+            Validator::withdraw_stake_from_pending_validator(validator, withdraw_amount, ctx);
+        } else {
+            let (found, validator_index) = find_validator(&self.active_validators, validator_address);
+            assert!(found, EVALIDATOR_NOT_FOUND);
+            let validator = Vector::borrow_mut(&mut self.active_validators, validator_index);
+            Validator::request_withdraw_stake_from_active_validator(validator, withdraw_amount);
+        }
+    }
+
+    /// Update the validator set at the end of epoch.
+    /// It does the following things:
+    ///   1. Distribute stake award.
+    ///   2. Process pending stake deposits and withdraws for each validator (`adjust_stake`).
+    ///   3. Process pending validator application and withdraws.
+    ///   4. At the end, we decide which validators will form the validator set for the next epoch.
+    public(friend) fun advance_epoch(
+        self: &mut ValidatorSet,
+        new_validator_set_size: u64,
+        new_min_validator_stake: u64,
+        new_max_validator_stake: u64,
+        ctx: &mut TxContext,
+    ) {
+        // TODO: Distribute stake rewards.
+
+        adjust_stake(&mut self.active_validators, ctx);
+
+        process_pending_withdraws(&mut self.active_validators, &mut self.pending_withdraws);
+        process_pending_validators(&mut self.active_validators, &mut self.pending_validators);
+
+        decide_new_active_validators(
+            &mut self.active_validators,
+            new_validator_set_size,
+            new_min_validator_stake,
+            new_max_validator_stake,
+        );
+    }
+
+    /// Checks whether a duplicate of `new_validator` is already in `validators`.
+    /// Two validators duplicate if they share the same sui_address or same IP or same name.
+    fun contains_duplicate_validator(validators: &vector<Validator>, new_validator: &Validator): bool {
+        let len = Vector::length(validators);
+        let i = 0;
+        while (i < len) {
+            let v = Vector::borrow(validators, i);
+            if (Validator::duplicates_with(v, new_validator)) {
+                return true
+            };
+            i = i + 1;
+        };
+        false
+    }
+
+    /// Find validator by `validator_address`, in `validators`.
+    /// Returns (true, index) if the validator is found, and the index is its index in the list.
+    /// If not found, returns (false, 0).
+    fun find_validator(validators: &vector<Validator>, validator_address: address): (bool, u64) {
```

Comment:
I think `Option<u64>` might be a bit safer, but not a big deal since this is an internal API.